test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java

```diff
@@ -0,0 +1,165 @@
+
+/**
+ * Represent a WebSocket Test connection.
+ */
+public class NanoWebSocketConnection extends NanoWSD.WebSocket {
+    private byte[] pingPayload;
+    private Scheduling scheduling;
+    private int closeInterval = 10000;
+    private int pingInterval;
+    private Stack<NanoWSD.WebSocketFrame> incomingMessages;
+    private Stack<NanoWSD.WebSocketFrame> outgoingMessages;
+    
+    public NanoWebSocketConnection(NanoHTTPD.IHTTPSession handshakeRequest){
+        super(handshakeRequest);
+        incomingMessages = new Stack<>();
+        outgoingMessages = new Stack<>();
+        scheduling = new Scheduling();
+    }
+    
+    /**
+     * Set up a ping Schedule.
+     * @param pingInterval how often in milliseconds send ping message.
+     * @param pingPayload the ping message.
+     */
+    public void setPingScheduling(int pingInterval, byte[] pingPayload){
+        this.pingPayload = pingPayload;
+        this.pingInterval = pingInterval;
+        scheduling.pingScheduling(pingInterval);
+    }
+    
+    /**
+     * Sends a ping message to the client. Ping payload defined by {@link NanoWebSocketConnection#setPingMessage(byte[])}
+     */
+    protected boolean ping(){
+        try {
+            super.ping(pingPayload);
+        } catch (IOException e) {
+            e.printStackTrace();
+            onClose(NanoWSD.WebSocketFrame.CloseCode.AbnormalClosure, "Connection Closed", false);
+            return false;
+        }
+        return true;
+    }
+    
+    @Override
+    protected void onOpen() {
+    }
+    
+    @Override
+    protected void onClose(NanoWSD.WebSocketFrame.CloseCode closeCode, String s, boolean b) { }
+    
+    @Override
+    protected void onMessage(NanoWSD.WebSocketFrame webSocketFrame) {
+        incomingMessages.push(webSocketFrame);
+    }
+    
+    @Override
+    protected void onPong(NanoWSD.WebSocketFrame webSocketFrame) {
+        incomingMessages.push(webSocketFrame);
+    }
+    
+    @Override
+    protected void onException(IOException e) { }
+    
+    public void setPingMessage(byte[] payload){
+        this.pingPayload = payload;
+    }
+    
+    public int getCloseInterval() {
+        return closeInterval;
+    }
+    
+    public void setCloseInterval(int closeInterval) {
+        this.closeInterval = closeInterval;
+    }
+    
+    public int getPingInterval() {
+        return pingInterval;
+    }
+    
+    public NanoWSD.WebSocketFrame popIncomingMessage(){
+        return incomingMessages.pop();
+    }
+    
+    /**
+     * Creates a schedule in order to send messages.
+     * @param messages Stack of the messages is going to be sent
+     * @param messageInterval the interval of the messages
+     */
+    public void setOutgoingMessageSchedule(Stack<NanoWSD.WebSocketFrame> messages, int messageInterval){
+        outgoingMessages = messages;
+        scheduling.messageScheduling(messageInterval);
+    }
+    
+    class SendPing extends TimerTask {
```

Comment:
The `TimerTask` is expected to be used with a `Timer`, cancelling this with the thread pool used has no effect (the executor needs to be shut down or the future cancelled).

Reply:
I introduced two Timer one for messages and one for pings. In addition, I introduced a boolean field which checks if the client response on ping message. If the client doesn't respond until the next scheduled ping, the connection will be closed.